File: api/prices/[prices].ts
/* eslint-disable import/no-anonymous-default-export */
import {
  lightFormat,
  parse,
  getUnixTime,
  isSameDay,
  addDays,
  subDays,
  isBefore,
  differenceInMinutes,
  differenceInDays,
  addHours,
} from "date-fns";
import { NowRequest, NowResponse } from "@vercel/node";
import get from "axios";

const AMBITO_API = "https://mercados.ambito.com";
const BLUELYTICS_API = "https://api.bluelytics.com.ar/v2/latest";
const COINGECKO_API = "https://api.coingecko.com/api/v3";
const ARS_MIN_DATE = parse("2002-01-11", "yyyy-MM-dd", new Date());
const BTC_MIN_DATE = parse("2013-04-30", "yyyy-MM-dd", new Date());

interface Price {
  date: number;
  value: number;
}

enum Pair {
  USDARS,
  USDBTC,
  USDSAT,
  ARSUSD,
  ARSBTC,
  ARSSAT,
  BTCUSD,
  BTCARS,
  SATUSD,
  SATARS,
}

export default async function (req: NowRequest, res: NowResponse) {
  const now = Date.now();
  const pairValue = req.query.prices.toString().toUpperCase();
  const pair = Pair[pairValue as keyof typeof Pair];
  let { from, to } = req.query;
  const f = from ? parseInt(from.toString()) : now;
  const t = to ? parseInt(to.toString()) : now;
  if (queryError(f, t, pair, res, now)) res.end();
  (async (p) => {
    switch (p) {
      case Pair.USDARS:
        return await getUsdArs(f, t);
      case Pair.USDBTC:
        return await getUsdBtc(f, t);
      case Pair.USDSAT:
        return await getUsdSat(f, t);
      case Pair.ARSUSD:
        return await getArsUsd(f, t);
      case Pair.ARSBTC:
        return await getArsBtc(f, t);
      case Pair.ARSSAT:
        return await getArsSat(f, t);
      case Pair.BTCUSD:
        return await getBtcUsd(f, t);
      case Pair.BTCARS:
        return await getBtcArs(f, t);
      case Pair.SATUSD:
        return await getSatUsd(f, t);
      case Pair.SATARS:
        return await getSatArs(f, t);
      default:
        res.status(404).json({ error: "Invalid pair" });
    }
  })(pair)
    .then((prices) => {
      if (prices !== undefined) {
        if (prices.length === 0)
          res.status(404).json({
            error: "Prices not found",
          });
        else res.json(prices);
      }
    })
    .catch((error) => {
      res.status(500).json({ error: error });
    });
}

function queryError(
  from: number,
  to: number,
  pair: Pair,
  res: NowResponse,
  now: number
) {
  if (pair === Pair.USDARS || pair === Pair.ARSUSD) {
    if (isBefore(from, ARS_MIN_DATE)) {
      res.status(404).json({
        error:
          "There is no data about Peso before " +
          lightFormat(ARS_MIN_DATE, "yyyy-MM-dd"),
      });
      return true;
    }
  } else if (isBefore(from, BTC_MIN_DATE)) {
    res.status(404).json({
      error:
        "There is no data about Bitcoin before " +
        lightFormat(BTC_MIN_DATE, "yyyy-MM-dd"),
    });
    return true;
  }
  if (isBefore(to, from)) {
    res.status(404).json({
      error: "Invalid date arguments",
    });
    return true;
  }
  if (isBefore(now, from) || isBefore(now, to)) {
    res.status(404).json({
      error: "There is no data about the future",
    });
    return true;
  }
  return false;
}

async function getUsdArs(from: number = Date.now(), to: number = Date.now()) {
  const now = Date.now();
  if (isSameDay(from, now) && isSameDay(to, now))
    return await get(BLUELYTICS_API).then((res: any) => {
      let price: Price = {
        date: now,
        value: parseFloat(res.data.blue.value_sell),
      };
      return [price];
    });
  else {
    let url = `${AMBITO_API}/dolar/informal/historico-general/${lightFormat(
      from,
      "dd-MM-yyyy"
    )}/${lightFormat(addDays(to, 1), "dd-MM-yyyy")}`;
    return await get(url)
      .then((res) => res.data)
      .then(async (res: any) => {
        res.shift(); // first item is a table header
        if (res.length === 0) return retry(from, to);
        let prices: Price[] = res
          .map((item: string[]) => {
            let price: Price = {
              date: addHours(
                parse(item[0], "dd-MM-yyyy", new Date()),
                3
              ).getTime(),
              value: parseFloat(item[2].replace(/,/g, ".")),
            };
            return price;
          })
          .reverse();
        if (!isSameDay(prices[0].date, from)) return retry(from, to);
        // looping for misinformation
        for (let index = 0; index < prices.length; index++) {
          const actual = prices[index];
          const last = prices.length - 1;
          if (index !== last) {
            let tomorrow = addDays(actual.date, 1).getTime();
            let next = prices[index + 1].date;
            // there are duplicated dates
            if (isSameDay(actual.date, next)) {
              prices.splice(index + 1, 1);
              index--;
              continue;
            }
            // fill dates without info with last price
            if (!isSameDay(tomorrow, next)) {
              let filledPrice: Price = {
                date: tomorrow,
                value: actual.value,
              };
              prices.splice(index + 1, 0, filledPrice);
            }
          } // same here
          else if (!isSameDay(prices[last].date, to)) {
            let filledPrice: Price = {
              date: addDays(prices[last].date, 1).getTime(),
              value: actual.value,
            };
            prices.splice(index + 1, 0, filledPrice);
          }
          if (prices[index].value === 0) {
            prices[index].value = prices[index - 1].value;
          }
        }
        // replace last filled price with today's one
        if (isSameDay(to, now)) {
          const today = await get(BLUELYTICS_API).then((res: any) => {
            let price: Price = {
              date: now,
              value: parseFloat(res.data.blue.value_sell),
            };
            return price;
          });
          prices[prices.length - 1] = today;
        }
        return prices;
      });
  }
}

async function retry(from: number, to: number) {
  let r: Price[] = await getUsdArs(subDays(from, 1).getTime(), to);
  r.shift();
  return r;
}

async function getBtcUsd(from: number = Date.now(), to: number = Date.now()) {
  const now = Date.now();
  const today = await get(
    `${COINGECKO_API}/simple/price?ids=bitcoin&vs_currencies=usd`
  ).then((res: any) => {
    let t: Price = {
      date: to,
      value: parseFloat(res.data.bitcoin.usd),
    };
    return t;
  });
  if (from === to || Math.abs(differenceInMinutes(from, now)) < 5) {
    return [today];
  } else {
    let url = `${COINGECKO_API}/coins/bitcoin/market_chart/range?vs_currency=usd&from=${getUnixTime(
      from
    )}&to=${getUnixTime(to)}`;
    return await get(url)
      .then((res: any) => res.data.prices)
      .then((res: number[][]) => {
        let prices = res.map((item: number[]) => {
          let price: Price = {
            date: item[0],
            value: item[1],
          };
          return price;
        });
        if (isSameDay(to, now) && Math.abs(differenceInDays(from, to)) > 90) {
          prices.push(today);
        }
        return prices;
      });
  }
}

async function getBtcArs(from: number = Date.now(), to: number = Date.now()) {
  const usdars = await getUsdArs(from, to);
  const btcusd = await getBtcUsd(from, to);
  let btcars: Price[] = [];
  let i = 0;
  for (const p of btcusd) {
    while (!isSameDay(usdars[i].date, p.date)) i++;
    let price: Price = {
      date: p.date,
      value: usdars[i].value * p.value,
    };
    btcars.push(price);
  }
  return btcars;
}

function mapValue(p: Price[], m: Function) {
  p.forEach((price) => {
    price.value = m(price.value);
  });
  return p;
}

function inverse(p: Price[]) {
  return mapValue(p, (v: number) => 1 / v);
}

function satoshi(p: Price[]) {
  return mapValue(p, (v: number) => v / Math.pow(10, 8));
}

async function getSatUsd(from: number = Date.now(), to: number = Date.now()) {
  return satoshi(await getBtcUsd(from, to));
}

async function getSatArs(from: number = Date.now(), to: number = Date.now()) {
  return satoshi(await getBtcArs(from, to));
}

async function getArsUsd(from: number = Date.now(), to: number = Date.now()) {
  return inverse(await getUsdArs(from, to));
}

async function getUsdBtc(from: number = Date.now(), to: number = Date.now()) {
  return inverse(await getBtcUsd(from, to));
}

async function getArsBtc(from: number = Date.now(), to: number = Date.now()) {
  return inverse(await getBtcArs(from, to));
}

async function getUsdSat(from: number = Date.now(), to: number = Date.now()) {
  return inverse(await getSatUsd(from, to));
}

async function getArsSat(from: number = Date.now(), to: number = Date.now()) {
  return inverse(await getSatArs(from, to));
}

File: api/prices/[prices].rs
use async_recursion::async_recursion;
use chrono::{Date, DateTime, Duration, NaiveDate, NaiveDateTime, Utc};
use http::{Response, StatusCode};
use now_lambda::{error::NowError, lambda, IntoResponse, Request};
use std::str::FromStr;
use std::vec;
use utils::*;

type Prices = Vec<Price>;
type Error = Box<dyn std::error::Error>;

fn main() -> Result<(), Error> {
    Ok(lambda!(handler))
}

#[tokio::main]
async fn handler(request: Request) -> Result<impl IntoResponse, NowError> {
    let now = Utc::now();
    let url = {
        let uri = request.uri().to_string();
        let url = reqwest::Url::parse(&uri);
        if url.is_err() {
            let err = HttpError::new(
                StatusCode::INTERNAL_SERVER_ERROR,
                "Internal server error",
                "The URL could not be parsed",
            );
            return Ok(err.res());
        }
        url.unwrap()
    };
    let pair: Pair = {
        let path = url.path_segments().unwrap().last().unwrap().to_uppercase();
        if path.len() != 6 {
            let err = HttpError::new(
                StatusCode::NOT_FOUND,
                "Invalid pair argument",
                "A pair must be six chars long",
            );
            return Ok(err.res());
        }
        let (base, quote) = {
            let res = path.split_at(3);
            let b = Coin::from_str(res.0);
            let q = Coin::from_str(res.1);
            if b.is_err() || q.is_err() {
                let detail = &format!("There is no data about {}/{}", res.0, res.1);
                let err = HttpError::new(StatusCode::NOT_FOUND, "Invalid pair argument", detail);
                return Ok(err.res());
            } else {
                (b.unwrap(), q.unwrap())
            }
        };
        (base, quote)
    };
    let mut query = url.query_pairs();

    let mut get_param = |param: &str| -> Result<DateTime<Utc>, Response<String>> {
        let err = {
            let detail = &format!("The '{}' date could not be parsed", param);
            let err = HttpError::new(
                StatusCode::BAD_REQUEST,
                "Invalid time range argument",
                detail,
            );
            err.res()
        };
        let res = query.find(|q| q.0 == param);
        if res.is_none() {
            Ok(now)
        } else {
            let t = res.unwrap().1.to_string();
            if t.len() == 0 {
                return Err(err);
            }
            let (secs, nsecs) = t.split_at(t.len() - 3);
            let secs = secs.parse::<i64>();
            let nsecs = nsecs.parse::<u32>();
            if secs.is_err() || nsecs.is_err() {
                return Err(err);
            }
            Ok(DateTime::<Utc>::from_utc(
                NaiveDateTime::from_timestamp(secs.unwrap(), nsecs.unwrap()),
                Utc,
            ))
        }
    };

    let from = match get_param("from") {
        Ok(param) => param,
        Err(err) => {
            return Ok(err);
        }
    };
    let to = match get_param("to") {
        Ok(param) => param,
        Err(err) => {
            return Ok(err);
        }
    };

    match pair {
        (Coin::USD, Coin::ARS) | (Coin::ARS, Coin::USD) => {
            if from.timestamp() < ARS_MIN_DATE {
                let err = HttpError::new(
                    StatusCode::NOT_FOUND,
                    "Invalid time range argument",
                    "There is no data about Argentine Peso before Januanry 11th 2002",
                );
                return Ok(err.res());
            }
        }
        _ => {
            if from.timestamp() < BTC_MIN_DATE {
                let err = HttpError::new(
                    StatusCode::NOT_FOUND,
                    "Invalid time range argument",
                    "There is no data about Bitcoin before April 30th 2013",
                );
                return Ok(err.res());
            }
        }
    }
    if to < from {
        let err = HttpError::new(
            StatusCode::BAD_REQUEST,
            "Invalid time range argument",
            "The 'from' data must be before 'to' date",
        );
        return Ok(err.res());
    }
    if now < from || now < to {
        let err = HttpError::new(
            StatusCode::BAD_REQUEST,
            "Invalid time range argument",
            "There is no data about the future",
        );
        return Ok(err.res());
    }
    let prices = match pair {
        (Coin::ARS, Coin::USD) => get_ars_usd(from, to).await,
        (Coin::ARS, Coin::BTC) => get_ars_btc(from, to).await,
        (Coin::ARS, Coin::SAT) => get_ars_sat(from, to).await,
        (Coin::USD, Coin::ARS) => get_usd_ars(from, to).await,
        (Coin::USD, Coin::BTC) => get_usd_btc(from, to).await,
        (Coin::USD, Coin::SAT) => get_usd_sat(from, to).await,
        (Coin::BTC, Coin::ARS) => get_btc_ars(from, to).await,
        (Coin::BTC, Coin::USD) => get_btc_usd(from, to).await,
        (Coin::SAT, Coin::ARS) => get_sat_ars(from, to).await,
        (Coin::SAT, Coin::USD) => get_sat_usd(from, to).await,
        _ => {
            let detail = &format!(
                "There is no data about {}/{}",
                pair.0.to_string(),
                pair.1.to_string()
            );
            let err = HttpError::new(StatusCode::NOT_FOUND, "Invalid pair argument", detail);
            return Ok(err.res());
        }
    };
    let response = match prices {
        Ok(p) => {
            let json = {
                let s = p
                    .iter()
                    .map(|i| PriceJson {
                        date: i.date.timestamp_millis(),
                        value: i.value,
                    })
                    .collect::<Vec<PriceJson>>();
                serde_json::to_string(&s).unwrap()
            };
            now_res(StatusCode::OK, json)
        }
        Err(err) => {
            let detail = &err.to_string();
            let err = HttpError::new(StatusCode::BAD_GATEWAY, "Bad Gateway", detail);
            err.res()
        }
    };

    Ok(response)
}

async fn get_usd_ars(from: DateTime<Utc>, to: DateTime<Utc>) -> Result<Prices, Error> {
    let now = Utc::now();
    let today = {
        let res = fetch(BLUELYTICS_API).await?;
        Price {
            date: now,
            value: res["blue"]["value_sell"]
                .as_f64()
                .ok_or(Error::from("Bluelytics parsing error"))?,
        }
    };
    if is_same_day(from, now) && is_same_day(to, now) {
        return Ok(vec![today]);
    } else {
        let url = format!(
            "{}/dolar/informal/historico-general/{}/{}",
            AMBITO_API,
            from.format("%d-%m-%Y"),
            to.format("%d-%m-%Y"),
        );
        let mut res = fetch(&url).await?;
        let res = res
            .as_array_mut()
            .ok_or(Error::from("Ámbito array parsing error"))?;
        res.remove(0);
        if res.len() == 0 {
            return retry(from, to).await;
        }
        res.reverse();

        let first_date = {
            let date = Date::<Utc>::from_utc(
                NaiveDate::parse_from_str(
                    res[0][0]
                        .as_str()
                        .ok_or(Error::from("Ámbito date parsing error"))?,
                    "%d-%m-%Y",
                )?,
                Utc,
            )
            .and_hms(0, 0, 0);
            let date = date + Duration::hours(3);
            date
        };
        if !is_same_day(first_date, from) {
            return retry(from, to).await;
        };
        let mut prices = res
            .iter()
            .map(|item| {
                let date = Date::<Utc>::from_utc(
                    NaiveDate::parse_from_str(
                        item[0]
                            .as_str()
                            .ok_or(Error::from("Ámbito date parsing error"))?,
                        "%d-%m-%Y",
                    )?,
                    Utc,
                )
                .and_hms(0, 0, 0);
                let date = date + Duration::hours(3); // from UTC−03:00 to UTC
                let value: f64 = item[2]
                    .as_str()
                    .ok_or(Error::from("Ámbito price parsing error"))?
                    .replace(",", ".")
                    .parse()?;
                Ok(Price { date, value })
            })
            .collect::<Result<Prices, Error>>()?;

        let mut index: usize = 0;
        // looping for misinformation
        while index < prices.len() {
            let actual = &prices[index];
            let last = prices.len() - 1;
            if index != last {
                let tomorrow = add_days(actual.date, 1);
                let next = prices[index + 1].date;
                // there are duplicated dates
                if is_same_day(actual.date, next) {
                    prices.remove(index + 1);
                    index -= 1;
                    continue;
                }
                // fill dates without info with last price
                if !is_same_day(tomorrow, next) {
                    let filled_price = Price {
                        date: tomorrow,
                        value: actual.value,
                    };
                    prices.insert(index + 1, filled_price);
                }
            }
            // same here
            else if !is_same_day(prices[last].date, to) {
                let filled_price = Price {
                    date: add_days(prices[last].date, 1),
                    value: actual.value,
                };
                prices.insert(index + 1, filled_price);
            }

            index += 1;
        }
        // replace last filled price with today's one
        if is_same_day(to, now) {
            prices.pop();
            prices.push(today);
        }
        return Ok(prices);
    }
}

#[async_recursion]
async fn retry(from: DateTime<Utc>, to: DateTime<Utc>) -> Result<Prices, Error> {
    return match get_usd_ars(sub_days(from, 1), to).await {
        Ok(mut r) => {
            r.remove(0);
            Ok(r)
        }
        Err(err) => Err(err),
    };
}

async fn get_btc_usd(from: DateTime<Utc>, to: DateTime<Utc>) -> Result<Prices, Error> {
    let now = Utc::now();
    let today = {
        let res = fetch(&format!(
            "{}/simple/price?ids=bitcoin&vs_currencies=usd",
            COINGECKO_API
        ))
        .await?;
        Price {
            date: to,
            value: res["bitcoin"]["usd"]
                .as_f64()
                .ok_or(Error::from("Coingecko price parsing error"))?,
        }
    };
    if difference_in_minutes(from, to).abs() < 5 && difference_in_minutes(from, now).abs() < 5 {
        return Ok(vec![today]);
    } else {
        let mut prices = {
            let url = format!(
                "{}/coins/bitcoin/market_chart/range?vs_currency=usd&from={}&to={}",
                COINGECKO_API,
                from.timestamp(),
                to.timestamp()
            );
            let res = fetch(&url).await?;
            let res = res["prices"]
                .as_array()
                .ok_or(Error::from("Coingecko array parsing error"))?;
            res.iter()
                .map(|item| {
                    let item = item
                        .as_array()
                        .ok_or(Error::from("Coingecko item parsing error"))?;
                    let secs = item[0]
                        .as_i64()
                        .ok_or(Error::from("Coingecko time parsing error"))?
                        / 1000;
                    let date = DateTime::from_utc(NaiveDateTime::from_timestamp(secs, 0), Utc);
                    let value = item[1]
                        .as_f64()
                        .ok_or(Error::from("Coingecko price parsing error"))?;
                    Ok(Price { date, value })
                })
                .collect::<Result<Prices, Error>>()?
        };
        if is_same_day(to, now) && difference_in_days(from, to).abs() > 90 {
            prices.push(today);
        }
        Ok(prices)
    }
}

async fn get_btc_ars(from: DateTime<Utc>, to: DateTime<Utc>) -> Result<Prices, Error> {
    let usd_ars = get_usd_ars(from, to).await?;
    let btc_usd = get_btc_usd(from, to).await?;
    let mut i: usize = 0;
    let btc_ars = btc_usd
        .iter()
        .map(|p| {
            while !is_same_day(usd_ars[i].date, p.date) && i < usd_ars.len() - 1 {
                i += 1;
            }
            let date = p.date;
            let value = usd_ars[i].value * p.value;
            Price { date, value }
        })
        .collect::<Prices>();
    Ok(btc_ars)
}

fn map_value(p: Prices, m: &dyn Fn(f64) -> f64) -> Prices {
    p.iter()
        .map(|i| Price {
            date: i.date,
            value: m(i.value),
        })
        .collect::<Prices>()
}

fn inverse(p: Prices) -> Prices {
    map_value(p, &|v: f64| v.recip())
}

fn satoshi(p: Prices) -> Prices {
    map_value(p, &|v: f64| v / 10_f64.powi(8))
}

async fn get_sat_usd(from: DateTime<Utc>, to: DateTime<Utc>) -> Result<Prices, Error> {
    let res = get_btc_usd(from, to).await?;
    Ok(satoshi(res))
}

async fn get_sat_ars(from: DateTime<Utc>, to: DateTime<Utc>) -> Result<Prices, Error> {
    let res = get_btc_ars(from, to).await?;
    Ok(satoshi(res))
}

async fn get_ars_usd(from: DateTime<Utc>, to: DateTime<Utc>) -> Result<Prices, Error> {
    let res = get_usd_ars(from, to).await?;
    Ok(inverse(res))
}

async fn get_usd_btc(from: DateTime<Utc>, to: DateTime<Utc>) -> Result<Prices, Error> {
    let res = get_btc_usd(from, to).await?;
    Ok(inverse(res))
}

async fn get_ars_btc(from: DateTime<Utc>, to: DateTime<Utc>) -> Result<Prices, Error> {
    let res = get_btc_ars(from, to).await?;
    Ok(inverse(res))
}

async fn get_usd_sat(from: DateTime<Utc>, to: DateTime<Utc>) -> Result<Prices, Error> {
    let res = get_sat_usd(from, to).await?;
    Ok(inverse(res))
}

async fn get_ars_sat(from: DateTime<Utc>, to: DateTime<Utc>) -> Result<Prices, Error> {
    let res = get_sat_ars(from, to).await?;
    Ok(inverse(res))
}
